--- libicu/uc/ucasemap.cpp	2019-12-23 14:38:39.764890499 +0300
+++ libicu/ucasemap.cpp	2020-01-27 19:42:13.816106977 +0300
@@ -687,13 +687,13 @@ void toUpper(uint32_t options,
             if (change) {
                 ByteSinkUtil::appendTwoBytes(upper, sink);
                 if ((data & HAS_EITHER_DIALYTIKA) != 0) {
-                    sink.Append(u8"\u0308", 2);  // restore or add a dialytika
+                    sink.Append("\u0308", 2);  // restore or add a dialytika
                 }
                 if (addTonos) {
-                    sink.Append(u8"\u0301", 2);
+                    sink.Append("\u0301", 2);
                 }
                 while (numYpogegrammeni > 0) {
-                    sink.Append(u8"\u0399", 2);
+                    sink.Append("\u0399", 2);
                     --numYpogegrammeni;
                 }
             }
