--- libicui18n/i18n/dtitvfmt.cpp	2019-12-23 14:38:40.214889289 +0300
+++ libicui18n/dtitvfmt.cpp	2020-07-21 14:35:34.435172232 +0300
@@ -232,7 +232,7 @@ DateIntervalFormat::operator==(const For
     {
         Mutex lock(&gFormatterMutex);
         if (fDateFormat != fmt->fDateFormat && (fDateFormat == NULL || fmt->fDateFormat == NULL)) {return FALSE;}
-        if (fDateFormat && fmt->fDateFormat && (*fDateFormat != *fmt->fDateFormat)) {return FALSE;}
+        if (fDateFormat && fmt->fDateFormat && !(*fDateFormat == *fmt->fDateFormat)) {return FALSE;}
     }
     // note: fFromCalendar and fToCalendar hold no persistent state, and therefore do not participate in operator ==.
     //       fDateFormat has the master calendar for the DateIntervalFormat.
