--- i18n/fmtable.cpp	2023-04-05 17:32:46.385668953 +0200
+++ fmtable.cpp	2023-04-06 14:01:47.787233628 +0200
@@ -56,7 +56,8 @@
 // Return TRUE if *a == *b.
 static inline UBool objectEquals(const UObject* a, const UObject* b) {
     // LATER: return *a == *b;
-    return *((const Measure*) a) == *((const Measure*) b);
+    //return *((const Measure*) a) == *((const Measure*) b);
+    return ((const Measure*) a)->operator== (*((const Measure*) b));
 }
 
 // Return a clone of *a.
