--- libicui18n/i18n/olsontz.cpp	2019-12-23 14:38:40.225889259 +0300
+++ libicui18n/olsontz.cpp	2020-07-21 14:14:11.005172691 +0300
@@ -624,7 +624,7 @@ OlsonTimeZone::hasSameRules(const TimeZo
     // be equal if their rules and transitions are equal
     if ((finalZone == NULL && z->finalZone != NULL)
         || (finalZone != NULL && z->finalZone == NULL)
-        || (finalZone != NULL && z->finalZone != NULL && *finalZone != *z->finalZone)) {
+        || (finalZone != NULL && z->finalZone != NULL && !(*finalZone == *z->finalZone))) {
         return FALSE;
     }
 
