--- libicui18n/i18n/stsearch.cpp	2019-12-23 14:38:40.234889235 +0300
+++ libicui18n/stsearch.cpp	2020-07-21 13:21:49.621459123 +0300
@@ -184,7 +184,7 @@ StringSearch::clone() const {
 // operator overloading ---------------------------------------------
 StringSearch & StringSearch::operator=(const StringSearch &that)
 {
-    if ((*this) != that) {
+    if (!((*this) == that)) {
         UErrorCode status = U_ZERO_ERROR;
         m_text_          = that.m_text_;
         m_breakiterator_ = that.m_breakiterator_;
