diff --git a/libicuuc/libicu/uc/uniset.cpp b/libicuuc/libicu/uc/uniset.cpp
index b73d612..d2d9653 100644
--- a/libicuuc/libicu/uc/uniset.cpp
+++ b/libicuuc/libicu/uc/uniset.cpp
@@ -302,7 +302,7 @@ UBool UnicodeSet::operator==(const UnicodeSet& o) const {
         if (list[i] != o.list[i]) return FALSE;
     }
     if (hasStrings() != o.hasStrings()) { return FALSE; }
-    if (hasStrings() && *strings != *o.strings) return FALSE;
+    if (hasStrings() && !(*strings == *o.strings)) return FALSE;
     return TRUE;
 }
 
